### PR TITLE
line:col positions in parser

### PR DESCRIPTION
**Pull request by kazcw**
### Pull Request Description

Add `line:column` information to source code references produced by the parser. This information will be used by GUI2 as part of the solution to #8134.

### Important Notes

- `parse_all_enso_files.sh` has been used to ensure this doesn't affect tree structures.
- `parse_all_enso_files.sh` now checks emitted locations for consistency, and has been used to verify that all line:col references match the values found by an independent scan of the source up to the given UTF8 position.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [ ] ~Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.~
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [] ~If GUI codebase was changed, the GUI was tested when built using `./run ide build`.~

### CODE REVIEW

File: lib/rust/parser/src/source/code.rs [hunk 1]
```diff
@@ -13,98 +13,133 @@ use crate::prelude::*;
...
+    pub utf8:  u32,
+    /// Offset from the beginning, in UTF-16 code units (two-byte words).
+    #[reflect(hide)]
+    pub utf16: u32,
+    /// Offset from the first line.
+    #[reflect(hide)]
+    pub line:  u32,
+    /// Offset from start of line, in Unicode characters.
+    #[reflect(hide)]
+    pub col:   u32,
```
Comment (Frizi) [inline lib/rust/parser/src/source/code.rs]: What does "unicode characters" mean here exactly? The amount of utf16 code units, or codepoints?

File: lib/rust/parser/src/source/code.rs [hunk 2]
```diff
@@ -13,98 +13,133 @@ use crate::prelude::*;
...
+pub struct Location {
+    /// Offset from the beginning, in UTF-8 code units (bytes).
+    #[reflect(hide)]
+    pub utf8:  u32,
+    /// Offset from the beginning, in UTF-16 code units (two-byte words).
+    #[reflect(hide)]
+    pub utf16: u32,
+    /// Offset from the first line.
+    #[reflect(hide)]
+    pub line:  u32,
```
Comment (Frizi) [inline lib/rust/parser/src/source/code.rs]: The "offset" is not clear enough here. Also it is not really clear what is a "line" really. In LSP, the lines are strictly defined to be separated by a numer of possible EOL characters. Therefore the file line ending style doesn't influence the locations at all. We should probably adapt exactly the same semantics. https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocuments

File: lib/rust/parser/debug/src/lib.rs [hunk 3]
```diff
@@ -23,12 +23,12 @@
 #![warn(unused_qualifications)]
 
 use enso_metamodel_lexpr::ToSExpr;
+use enso_parser::source::code::debug::LocationCheck;
 use enso_reflect::Reflect;
 use lexpr::Value;
 use std::collections::HashSet;
 
 
-
```
Comment (farmaazon) [inline lib/rust/parser/debug/src/lib.rs]: Keep this newline.

File: lib/rust/parser/debug/src/lib.rs [hunk 4]
```diff
@@ -122,10 +122,18 @@ fn strip_hidden_fields(tree: Value) -> Value {
         ":spanLeftOffsetVisible",
         ":spanLeftOffsetCodeReprBegin",
         ":spanLeftOffsetCodeReprLen",
-        ":spanLeftOffsetCodeUtf16",
+        ":spanLeftOffsetCodeLenUtf8",
+        ":spanLeftOffsetCodeLenUtf16",
+        ":spanLeftOffsetCodeLenNewlines",
+        ":spanLeftOffsetCodeLenLineChars16",
```
Comment (farmaazon) [inline lib/rust/parser/debug/src/lib.rs:129]: If the node has multiple lines, this is the length of the last line?

File: lib/rust/parser/src/source/code.rs [hunk 5]
```diff
@@ -13,98 +13,133 @@ use crate::prelude::*;
...
     #[reflect(as = "crate::serialization::Code", flatten, hide)]
     #[deref]
-    pub repr:     StrRef<'s>,
-    #[reflect(hide)]
-    offset_utf16: u32,
-    #[reflect(hide)]
-    utf16:        u32,
+    pub repr: StrRef<'s>,
+    #[reflect(flatten)]
+    offset:   Location,
```
Comment (farmaazon) [inline lib/rust/parser/src/source/code.rs]: I would rename the field as well, as it's now more than just the offset.

File: lib/rust/parser/src/source/code.rs [hunk 6]
```diff
@@ -205,17 +240,37 @@ impl<'s> AddAssign<&Code<'s>> for Code<'s> {
...
+        let mut newlines = 0;
+        let mut line_chars16 = 0;
+        for c in s.chars() {
+            let char_len16 = c.len_utf16() as u32;
+            utf16 += char_len16;
+            line_chars16 += char_len16;
+            if c == '\n' {
+                newlines += 1;
+                line_chars16 = 0;
+            }
```
Comment (farmaazon) [inline lib/rust/parser/src/source/code.rs]: Somewhere else we say about interpreting CRLF, CR and LF as newlines, but here we consider only LF.

**Review by farmaazon — COMMENTED** [review]
I'll check lexer.rs later today.

File: lib/rust/parser/src/source/code.rs (hunk 1 above)
Comment (kazcw) [reply lib/rust/parser/src/source/code.rs]: It should be UTF16 code units. Fixed.

File: lib/rust/parser/src/source/code.rs (hunk 2 above)
Comment (kazcw) [reply lib/rust/parser/src/source/code.rs]: Fixed it to match LSP (which matches the rest of the language syntax).

File: lib/rust/parser/debug/src/lib.rs (hunk 4 above)
Comment (kazcw) [reply lib/rust/parser/debug/src/lib.rs:129]: Yes, a line:col length contains the information necessary to reach an end line:col location from a start location.

File: lib/rust/parser/src/source/code.rs (hunk 5 above)
Comment (kazcw) [reply lib/rust/parser/src/source/code.rs]: We were overusing "offset"; renamed it to "start".

File: lib/rust/parser/src/source/code.rs (hunk 6 above)
Comment (kazcw) [reply lib/rust/parser/src/source/code.rs]: Fixed and added testing.

File: lib/rust/parser/src/lexer.rs [hunk 7]
```diff
@@ -1831,6 +1857,39 @@ mod tests {
...
+    #[test]
+    fn test_based_numbers() {
+        lex_and_validate_spans("0x23");
+        lex_and_validate_spans("2_010101");
+    }
+
+    #[test]
+    fn test_line_endings() {
+        //lex_and_validate_spans("Windows\r\n...");
+        //lex_and_validate_spans("Linux\n...");
```
Comment (farmaazon) [inline lib/rust/parser/src/lexer.rs]: Why is it commented?

**Review by farmaazon — APPROVED** [review]

**Review by JaroslavTulach — COMMENTED** [review]
The Linux engine job is OK, but Mac and Windows are failing:
- [Linux](https://github.com/enso-org/enso/actions/runs/6746357488/job/18340115728?pr=8203) - OK
- [Mac](https://github.com/enso-org/enso/actions/runs/6746357488/job/18340115145?pr=8203#step:10:4635) - probably line endings
- [Windows](https://github.com/enso-org/enso/actions/runs/6746357488/job/18340116544?pr=8203#step:10:828) panic in native (probably Rust) code

File: lib/rust/parser/src/lexer.rs (hunk 7 above)
Comment (kazcw) [reply lib/rust/parser/src/lexer.rs]: Fixed

**Comment by farmaazon** [comment]
What is the status of this PR? Are the @Frizi comments addressed? Do we want to merge it or discard in favor of #8236 ?

**Comment by kazcw** [comment]
I think it's an improvement regardless of #8236. In the future we might make use of it for a more edit-resilient metadata map format.

**Review by Frizi — APPROVED** [review]

**Pull request merged**